google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/internal/AckSetTrackerImpl.java

```diff
@@ -57,23 +99,26 @@ public AckSetTrackerImpl(Committer committer) throws ApiException {
   public Runnable track(SequencedMessage message) throws CheckedApiException {
     final Offset messageOffset = message.offset();
     try (CloseableMonitor.Hold h = monitor.enter()) {
-      checkArgument(receipts.isEmpty() || receipts.peekLast().value() < messageOffset.value());
-      receipts.addLast(messageOffset);
+      checkArgument(
+          receipts.isEmpty() || receipts.peekLast().offset.value() < messageOffset.value());
+      Receipt receipt = new Receipt(messageOffset, this);
+      receipts.addLast(receipt);
+      return receipt::onAck;
     }
-    return new Runnable() {
-      private final AtomicBoolean wasAcked = new AtomicBoolean(false);
+  }
 
-      @Override
-      public void run() {
-        if (wasAcked.getAndSet(true)) {
-          CheckedApiException e =
-              new CheckedApiException("Duplicate acks are not allowed.", Code.FAILED_PRECONDITION);
-          onPermanentError(e);
-          throw e.underlying;
-        }
-        onAck(messageOffset);
-      }
-    };
+  @Override
+  public void waitUntilCommitted() throws CheckedApiException {
+    List<Receipt> receiptsCopy;
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      receiptsCopy = ImmutableList.copyOf(receipts);
+    }
+    receiptsCopy.forEach(Receipt::clear);
```

Comment:
nit. add a comment that we're doing this to avoid lock order problems

Reply:
Done